infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -173,226 +182,265 @@ public void handleTailRequest(CorfuPayloadMsg<TailsRequest> msg, ChannelHandlerC
-        r.sendResponse(ctx, msg, CorfuMsgType.TRIM_MARK_RESPONSE.payloadMsg(streamLog.getTrimMark()));
+    @RequestHandler(type = PayloadCase.TRIM_MARK_REQUEST)
+    public void handleTrimMarkRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        if (log.isTraceEnabled()) {
+            log.trace("handleTrimMarkRequest: received a trim mark request {}", TextFormat.shortDebugString(req));
+        }
+
+        // Note: we reuse the request header as the ignore_cluster_id and
+        // ignore_epoch fields are the same in both cases.
+        r.sendResponse(getResponseMsg(req.getHeader(), getTrimMarkResponseMsg(streamLog.getTrimMark())), ctx);
     }
 
     /**
      * Service an incoming query for the committed tail on this log unit server.
      */
-    @ServerHandler(type = CorfuMsgType.COMMITTED_TAIL_REQUEST)
-    public void handleCommittedTailRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        log.trace("handleCommittedTailRequest: received a committed log tail request {}", msg);
-        r.sendResponse(ctx, msg, CorfuMsgType.COMMITTED_TAIL_RESPONSE.payloadMsg(streamLog.getCommittedTail()));
+    @RequestHandler(type = PayloadCase.COMMITTED_TAIL_REQUEST)
+    public void handleCommittedTailRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        if (log.isTraceEnabled()) {
+            log.trace("handleCommittedTailRequest: received a "
+                    + "committed log tail request {}", TextFormat.shortDebugString(req));
+        }
+
+        // Note: we reuse the request header as the ignore_cluster_id and
+        // ignore_epoch fields are the same in both cases.
+        r.sendResponse(getResponseMsg(req.getHeader(),
+                getCommittedTailResponseMsg(streamLog.getCommittedTail())), ctx);
     }
 
     /**
      * Service an incoming request to update the current committed tail.
      */
-    @ServerHandler(type = CorfuMsgType.UPDATE_COMMITTED_TAIL)
-    public void updateCommittedTail(CorfuPayloadMsg<Long> msg,
-                                    ChannelHandlerContext ctx, IServerRouter r) {
-        log.trace("updateCommittedTail: received request to update committed tail {}", msg);
-        streamLog.updateCommittedTail(msg.getPayload());
-        r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
+    @RequestHandler(type = PayloadCase.UPDATE_COMMITTED_TAIL_REQUEST)
+    public void handleUpdateCommittedTailRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        if (log.isTraceEnabled()) {
+            log.trace("handleUpdateCommittedTailRequest: received request to "
+                    + "update committed tail {}", TextFormat.shortDebugString(req));
+        }
+
+        streamLog.updateCommittedTail(req.getPayload().getUpdateCommittedTailRequest().getCommittedTail());
+        HeaderMsg responseHeader = getHeaderMsg(req.getHeader(), false, true);
```

Comment:
I actually think that this method does nor provide much value. The problem I have with it is that arguments two and three are not type safe, so it is pretty easy to mess us the ordering. When using the `Builder` directly, it is much harder, since the parameter names are referenced explicitly.

Reply:
Alternatively, you can defines types for each of the paramets:

```
    public enum ClusterId {
        CHECK,
        IGNORE
    }
    
    public enum Epoch {
        CHECK,
        IGNORE
    }
```
And change you method implementation:
```
    public static HeaderMsg getHeaderMsg(HeaderMsg header, ClusterId ignoreClusterId, Epoch ignoreEpoch) {
        return HeaderMsg.newBuilder()
                .mergeFrom(header)
                .setIgnoreClusterId(ignoreClusterId == ClusterId.IGNORE)
                .setIgnoreEpoch(ignoreEpoch == Epoch.IGNORE)
                .build();
    }
```
So a call to that function would become: `getHeaderMsg(req.getHeader(), ClusterId.IGNORE, Epoch.IGNORE);`

Reply:
I like this approach, but as discussed, changing the `getHeaderMsg` method signatures requires updating all of the servers, clients and tests. I'll address this in a different PR.